Kentico/kontent-delivery-sdk-net
# Distributed cache + Async API

### Motivation

Fixes 
- #213  - IContentLinkUrlResolver not async friendly (recommended reading)
- #196 - Add support for IDistributedCache to the SDK - approach no. 1 (recommended reading)
- #222 - Add performance tests

Follows up on #218 (also a recommended reading for understanding the context)

### Checklist

- [ ] Docs have been updated (if applicable) -> will be

### Breaking changes, features & upgrade guide (for the whole v14 release)
- all models have their interfaces extracted to `Kentico.Kontent.Delivery.Abstractions` and we only return interfaces
  - for Asset we have IAsset, for ContentType there is an IContentType, etc.
- `DeliveryCacheManager` was renamed to `MemoryCacheManager`
- there is a new implementation of `IDeliveryCacheManager` called `DistributedCacheManager` which implements the `IDistributedCache` interface using [BSON serialization](http://bsonspec.org/)
  - it's possible to register the cache using `Kentico.Kontent.Delivery.Caching.Extensions.ServiceCollectionExtensions.AddDeliveryClientCache()` by changing `DeliveryCacheOptions.CacheType` from `Memory` to `Distributed`. by default, it registers the `MemoryDistributedCache`. if you want to use a different implementation (e.g. redis, you need to register its instance before calling `AddDeliveryClientCache()`
- `IPropertyValueConverter.GetPropertyValue` now is now strongly typed. Instead of `JToken`, you receive `ContentElementValue<T>` with Name, Codename, Type, and Value properties where Value is of type T and T is the type of your property (DateTime, string, int, Asset....). also, the method is async now.
- methods in `IContentLinkUrlResolver`, `IModelProvider`, `IInlineContentItemsProcessor` are now async, their signatures remain the same
- some models are now more specific (contain e.g. `Guid` instead of `string` where it was appropriate)
- places which returned `ContentElement` now return `IContentElement`. plus, based on the type of the element, they can return a type castable to `IMultipleChoiceElement` or `ITaxonomyElement` to allow strongly typed access to members specific to these types
- the `IDeliveryClient` now contains only async methods that operate upon strongly-typed models. all JSON-based methods were removed. if someone wishes to access the raw json, all `Delivery*Response` objects contain an object called ApiResponse of the `IApiResponse` type. this property contains low-level data like `string Content`, `string RequestUrl`, or `string ContinuationToken`.


### Notes for the reviewer
- we got rid of the static `JsonConvert` and we inject a custom serializer instance via DI (see `services.TryAddSingleton<JsonSerializer>(new DeliveryJsonSerializer());` in `ServiceCollectionExtensions`
- this custom instance - `DeliveryJsonSerializer` - facilitates the instantiation of concrete types for interface type models via `DeliveryContractResolver` and `DeliveryServiceCollection`. it also facilitates translation to the best-suited models for content elements (see the `ContentElementConverter`)
- all concrete model types have been marked as internal
  - their deserialization was unified. we now only use `.ToObject<T>()` and don't use anything `JToken` or `JObject`-related. a lot of code was relying on JTokens, now we use a strongly-typed approach. also, all model constructors were unified to allow this kind of deserialization.
- one of such places was `RichTextContentConverter` and all kinds of `IPropertyValueConverter`s
- getting rid of Newtonsoft.Json types from models and interfaces allowed to free the `Kentico.Kontent.Delivery.Abstractions` from all non-framework dependencies. also, note that all types reside in the main namespace. this makes it easy to use them anywhere without injecting multiple using statements.
- `IReadOnlyList` occurrences were changed to `IList` to allow (de)serialization of the whole Delivery*Response types (the same goes with  `IReadOnlyDictionary` and `IDictionary`)
- distributed caching implementation details:
  - the BSON serialization loses precision for DateTime formats, therefore unit tests are decorated with `AssertionExtensions.DateTimesBsonCorrection` which allows 1ms difference
  - also, the implementation of `IDistributedCache` is much simpler than the other one. so it doesn't allow invalidating by cache dependencies, only by cache keys. this is reflected in the unit tests using `[InlineData(CacheTypeEnum.Memory)]` and `[InlineData(CacheTypeEnum.Distributed)]` for the given theories
  - for tests, we use the `MemoryDistributedCache`
- another type that is now being injected via DI is the `IHtmlParser`. creating an instance of this type is quite costly, so it should save some resources
- the `ContentLinkResolver` has been refactored to support async
- the Delivery*Response model types have been freed of all logic
- the `InlineContentItemsProcessor` now relies on `ParseDocumentAsync` instead of its sync `Parse` counterpart
- the `Content` property in the `ApiResponse` allows access to low-level data. it might make sense to switch it to an async method. any thoughts?

### Performance tests
#### Code benchmarks
- there is a new project called `Kentico.Kontent.Delivery.Benchmarks`. you can run the benchmarks by switching to the "Release" build configuration and running the respective unit tests or by compiling in the Release mode and running the resulting exe. the results of the benchmarks are in the 'Kentico.Kontent.Delivery.Benchmarks\bin\Release\netcoreapp3.1\BenchmarkDotNet.Artifacts' folder

**Results:**
v13.0.1:
![Benchmark Net 13 0 1](https://user-images.githubusercontent.com/9810625/90247004-03b4c400-de36-11ea-963e-13da775a4cbd.png)

v14.0.0-beta12:
![Benchmark Net 14 0 0-beta12](https://user-images.githubusercontent.com/9810625/90247009-06171e00-de36-11ea-9974-1c99dde848ee.png)

There is a notable performance drop when it comes to the SDK code-only benchmark. My suspicion is that there is some overhead related to `async`. We might want to profile and investigate the exact causes further.

#### Load tests
Configuration: Release
Host: Kestrel
Tool: [Bombardier](https://github.com/codesenberg/bombardier)
Site: [Kentico Kontent .NET Boilerplate](https://github.com/Kentico/kontent-boilerplate-net)
LogLevel: Warning
Caching: off

PS1: `.\bombardier.exe -c 128  -d 10s -l https://localhost:5001`
PS2: `dotnet counters monitor --process-id <PID> System.Runtime Microsoft.AspNetCore.Hosting`

v13.0.1:
![10s_13](https://user-images.githubusercontent.com/9810625/90247399-c4d33e00-de36-11ea-8cd9-5f0569c87926.png)

v14.0.0-beta12:
![10s_14](https://user-images.githubusercontent.com/9810625/90247402-c56bd480-de36-11ea-8ce9-7b9043453f09.png)

The throughput and latency distribution of both versions are comparable and within the margin of error. So this good news.

#### Conclusions & suggestions
My suggestion is to release an RC version of the SDK and let users test it in real-world scenarios. In the meantime, we can profile the code and run more sophisticated, composite benchmarks in isolated environments and keep improving the code.

### What to verify
- test together with the new generator: https://github.com/Kentico/kontent-generators-net/releases/tag/6.0.0-beta4
- try migrating an existing app (own or https://github.com/Kentico/kontent-sample-app-razorpages or https://github.com/Kentico/kontent-boilerplate-net)
- test against the live API
- verify the list of breaking changes 
- verify the perf tests (see the previous chapter)
- verify that the developer experience hasn't been impacted negatively
